### ISSUE TITLE
provide a way for applications to request loader protocols

### ISSUE THREAD

**Issue by mmerickel** [Pylons/plaster#6]
Ideally `plaster` could ship with a default loader for INI files very easily via the stdlib `ConfigParser`, and a `plaster_yaml` and `plaster_toml` could also be written for `.yml` and `.toml` file extensions respectively. Then `plaster` would be immediately usable for applications that just want to abstract the file format. 

However, right now, `plaster.get_loader` will get upset if more than one loader is defined for the same file format even if the file format is not what's interesting to the loader. For example, with `plaster_pastedeploy` we are adding extra methods to the loader to load wsgi-specific constructs such as `loader.get_wsgi_app`. Consider this a loader that supports a "wsgi" loader protocol (effectively it defines a loader with extra methods).

Pyramid would then only care about loaders which supported this protocol. A regular `.ini` loader would be unseen.

``` python
loader = plaster.get_loader('development.ini', protocols=['wsgi'])
```

Currently, `plaster` leaves the loading completely up to the user to specify the appropriate loader. For example, if it's ambiguous then a user needs to say `ini+pastedeploy://development.ini`. The goal here is to hide this from the user and let them continue with `development.ini` while the application can help disambiguate by specifying the desired feature-set of the loader. The user could always override this by specifying the exact loader like they can now. This is strictly about reducing the number of cases in which they must do this in order to make the syntax easily digestible.

So far I don't have a solid way to do this, but my first thought is to add an extra bit of info to the entry point registration. Alternatively, I could further specify the `ILoaderFactory` that addons provide (usually just a `Loader(uri)` class right now such that we can query it at runtime about its supported protocols. This is less efficient because it requires loading the code behind every entry point instead of using simply the entry point name.

**Comment by mmerickel** [comment]
I think one option is just to treat the thing on the right side of the `+` as the protocol. I think this works out fairly seamlessly. We could either support multi-level protocols (multiple `+` delimiters) or just one. Assuming we just support one then:

``` python
get_loader('development.ini', protocol='wsgi') # -> ini+wsgi
get_loader('ini+pastedeploy://development.ini', protocol='wsgi') # -> ini+pastedeploy
```

Assuming multiple delimiters:

``` python
get_loader('development.ini', protocol='wsgi') # -> ini+wsgi
get_loader('ini+pastedeploy://development.ini', protocol='wsgi') # -> ini+pastedeploy+wsgi
```

I'm leaning toward the first option myself which allows the user to override the selected loader explicitly.

**Comment by mmerickel** [comment]
I'm in a bit of a rut thinking about this. I'm not sure if it's really a solvable problem given the ambiguity in using the file extension. The issue is that there's so many different formats for an ini file. Only part of it is the parser and the rest is the higher-level constructs built on top of the format which are left undefined and ambiguous. For example, Python 3's configparser supports several different interpolation mechanisms including zc.buildout-style `${section:item}` and legacy `%(default_item)` (which is what PasteDeploy uses). The decision about which one to use is going to really matter to the user who wrote the ini file. In plaster's api this decision is only relevant to the user, not the app. However the app will care about the api exposed by the loader (whether it supports `get_wsgi_app` or not).

I think the best we can do is have the application require a particular feature-set on the loader and if the ini file fails to parse properly then the user will need to be more explicit about which loader to use by specifying the protocol.

Here's the use case I'm hung up on:
- Pyramid wants a loader that supports wsgi features (`get_wsgi_app`).
- CLI scripts I write don't care.
- I want to share the config file between my scripts and Pyramid.
- If `plaster` ships its own ini parser, and Pyramid installs `plaster_pastedeploy` then we now have two ini parsers available.
- Above I discussed how Pyramid would ensure that it always gets `plaster_pastedeploy` because it would request the loader that supports wsgi features.
- It would be silly for the CLI scripts to specifically request a loader with wsgi features.
- How does the CLI get the same loader as the Pyramid app such that the file parses the exact same way?

Part of the issue here is the sort order / priority of the entry points I think. For example, maybe plaster's ini support could specify itself as a fallback such that it isn't used unless there is no other option. This may work but has its own problems and doesn't scale past a single fallback.

Anyway, as I said I'm in a rut with this and it's the primary blocker from really endorsing this library.

**Comment by mmerickel** [comment]
I've implemented a possible solution in #8.
